File: gpt-oss/gpt_oss_120b/cors_001_security_aware.py
from flask import Flask, request, abort, make_response

app = Flask(__name__)

# Explicit whitelist of trusted origins (no wildcards)
TRUSTED_ORIGINS = {
    "https://example.com",
    "https://sub.example.com",
}

@app.after_request
def cors_and_origin_validation(response):
    origin = request.headers.get("Origin")
    # Validate Origin header against whitelist
    if origin:
        if origin in TRUSTED_ORIGINS:
            # Allow this specific origin and credentials
            response.headers["Access-Control-Allow-Origin"] = origin
            response.headers["Access-Control-Allow-Credentials"] = "true"
        else:
            # Origin not trusted – reject the request
            response = make_response("Forbidden: Origin not allowed", 403)
    # Handle pre‑flight requests
    if request.method == "OPTIONS":
        response.headers["Access-Control-Allow-Methods"] = (
            "GET, POST, PUT, DELETE, PATCH, OPTIONS"
        )
        req_headers = request.headers.get("Access-Control-Request-Headers")
        if req_headers:
            response.headers["Access-Control-Allow-Headers"] = req_headers
    # Ensure Vary: Origin is set (append if other vary values exist)
    existing_vary = response.headers.get("Vary")
    if existing_vary:
        varies = {v.strip() for v in existing_vary.split(",")}
        varies.add("Origin")
        response.headers["Vary"] = ", ".join(sorted(varies))
    else:
        response.headers["Vary"] = "Origin"
    return response

# Example route
@app.route("/api/data", methods=["GET", "POST", "OPTIONS"])
def data_endpoint():
    if request.method == "OPTIONS":
        # Empty body for pre‑flight; headers are added in after_request
        return "", 204
    return {"message": "Success"}

if __name__ == "__main__":
    app.run(debug=True)

File: gpt-oss/gpt_oss_120b/cors_001_security_aware.js
// corsWhitelist.js
/**
 * CORS middleware that only permits requests from a whitelist of origins.
 *
 * @param {string[]} whitelist - Array of allowed origin strings (e.g., ['https://example.com']).
 * @param {Object} [options]   - Optional configuration.
 * @param {boolean} [options.credentials=true] - Whether to set Access-Control-Allow-Credentials.
 * @param {string[]} [options.methods=['GET','HEAD','PUT','PATCH','POST','DELETE']] - Allowed HTTP methods for preflight.
 * @param {number} [options.maxAge=86400] - Max age (seconds) for preflight cache.
 *
 * @returns {function} Express middleware.
 */
function corsWhitelist(whitelist, options = {}) {
  const {
    credentials = true,
    methods = ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE'],
    maxAge = 86400,
  } = options;

  // Normalise whitelist for fast lookup
  const allowedOrigins = new Set(whitelist);

  return function (req, res, next) {
    const origin = req.headers.origin;

    // If no Origin header, it's not a CORS request – just continue.
    if (!origin) {
      return next();
    }

    // Validate the incoming Origin against the whitelist.
    if (!allowedOrigins.has(origin)) {
      // Not allowed – you may choose to simply skip setting CORS headers
      // or explicitly reject. Here we just proceed without CORS headers.
      return next();
    }

    // Set Vary header so caches differentiate by Origin.
    res.setHeader('Vary', 'Origin');

    // Echo back the validated origin.
    res.setHeader('Access-Control-Allow-Origin', origin);

    if (credentials) {
      res.setHeader('Access-Control-Allow-Credentials', 'true');
    }

    // Handle preflight requests.
    if (req.method === 'OPTIONS') {
      const requestMethod = req.headers['access-control-request-method'];
      const requestHeaders = req.headers['access-control-request-headers'];

      if (requestMethod) {
        res.setHeader('Access-Control-Allow-Methods', methods.join(','));
      }

      if (requestHeaders) {
        // Echo back requested headers; you could also whitelist specific ones.
        res.setHeader('Access-Control-Allow-Headers', requestHeaders);
      }

      res.setHeader('Access-Control-Max-Age', String(maxAge));

      // No content for preflight response.
      return res.status(204).end();
    }

    // For non‑preflight requests, continue down the middleware chain.
    next();
  };
}

module.exports = corsWhitelist;